### ISSUE TITLE
Connection timed out : when using python -m visdom.server

### ISSUE THREAD

**Issue by huoliangyu** [facebookresearch/visdom#185]
Hi, when I use 
`python -m visdom.server`
to start visdom server, I got Connection timed out error as following:
`Traceback (most recent call last):
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/urllib/request.py", line 1318, in do_open
    encode_chunked=req.has_header('Transfer-encoding'))
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/http/client.py", line 964, in send
    self.connect()
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/http/client.py", line 1392, in connect
    super().connect()
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/http/client.py", line 936, in connect
    (self.host,self.port), self.timeout, self.source_address)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/socket.py", line 722, in create_connection
    raise err
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/socket.py", line 713, in create_connection
    sock.connect(sa)
TimeoutError: [Errno 110] Connection timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/site-packages/visdom/server.py", line 615, in download_scripts
    data = opener.open(req).read()
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/urllib/request.py", line 526, in open
    response = self._open(req, data)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/urllib/request.py", line 544, in _open
    '_open', req)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/urllib/request.py", line 1361, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/urllib/request.py", line 1320, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno 110] Connection timed out>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/site-packages/visdom/server.py", line 636, in <module>
    download_scripts()
  File "/root/anaconda2/envs/gtn_env/lib/python3.6/site-packages/visdom/server.py", line 619, in download_scripts
    logging.error('Error {} while downloading {}'.format(exc.code, key))
AttributeError: 'URLError' object has no attribute 'code'
`
I am a new one for pytorch and visdom. I would like to use visdom for visualization. I use anaconda2, python3.6 as default setting. What should I do? Thanks a lot !

**Comment by lvdmaaten** [comment]
Visdom downloads some external JS dependencies at server start-up time. It looks like those downloads fail for you. Do you need a proxy to be able to access public URLs from your machine?

**Comment by huoliangyu** [comment]
Thanks for your suggestion . However,  It seems that I don' t need a proxy, since another machine is able to run the visdom server normally with the same Internet setting as mine. Is there any possible for other problems ? Thanks a lot.

**Comment by JackUrb** [comment]
Still, if your `urllib` module is up to date it would seem at least one of the packages must be failing, though it's hard to say which one. Can you list the contents of your `/root/anaconda2/envs/gtn_env/lib/python3.6/site-packages/visdom/static` directory and sub directories?

**Comment by huoliangyu** [comment]
Thanks for reply. The directory  "............../static" contains :
`css  fonts  index.html  js`
The directory  "............../static/js" contains:
`bootstrap.min.js  jquery.min.js  main.js  react-dom.min.js  react-react.min.js`
Does it  help ?

**Comment by JackUrb** [comment]
Is there anything at all in fonts or css? If not it would appear the first thing missing in the order of things downloaded belongs in the font folder, namely the script downloaded by visdom from `https://unpkg.com/classnames@2.2.5` and stored as `visdom/static/fonts/classnames`. If nothing is in your fonts directory can you download that file manually, save it as classnames, and put it in fonts before continuing? Then see if more files populate?

**Comment by huoliangyu** [comment]
I have checked the `static/fonts` directory. It contains:
`classnames  layout_bin_packer`
In "classnames" , it is  commented at top as following:
`  Copyright (c) 2016 Jed Watson.
  Licensed under the MIT License (MIT), see
  http://jedwatson.github.io/classnames`
I am not sure if it is dowloaded from `https://unpkg.com/classnames@2.2.5`.
Do I  still need to download that file manually as you say ?

**Comment by JackUrb** [comment]
No that's the correct file. It seems all the unpkg.com files from the first batch download then. Based on what you've listed the first file you're getting stuck on is `https://cdn.rawgit.com/STRML/react-grid-layout/0.14.0/dist/react-grid-layout.min.js`, which should be saved as `visdom/static/js/react-grid-layout.min.js`. Do you have this file?

**Comment by JackUrb** [comment]
Correct, then try to start the server again and see if it makes it to running. In general as you run into problems with CDNs, the order that visdom downloads and saves files is as follows:
```
download from https://unpkg.com/jquery@3.1.1/dist/jquery.min.js 
and put at visdom/static/js/jquery.min.js
download from https://unpkg.com/bootstrap@3.3.7/dist/js/bootstrap.min.js 
and put at visdom/static/js/bootstrap.min.js
download from https://unpkg.com/react-resizable@1.4.6/css/styles.css 
and put at visdom/static/css/react-resizable-styles.css
download from https://unpkg.com/react-grid-layout@0.14.0/css/styles.css
and put at visdom/static/css/react-grid-layout-styles.css
download from https://unpkg.com/react@15.6.1/dist/react.min.js
and put at visdom/static/js/react-react.min.js
download from https://unpkg.com/react-dom@15.6.1/dist/react-dom.min.js 
and put at visdom/static/js/react-dom.min.js
download from https://unpkg.com/classnames@2.2.5 
and put at visdom/static/fonts/classnames
download from https://unpkg.com/layout-bin-packer@1.2.2 
and put at visdom/static/fonts/layout_bin_packer
download from https://cdn.rawgit.com/STRML/react-grid-layout/0.14.0/dist/react-grid-layout.min.js 
and put at visdom/static/js/react-grid-layout.min.js
download from https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_SVG 
and put at visdom/static/js/mathjax-MathJax.js
download from https://cdn.rawgit.com/plotly/plotly.js/master/dist/plotly.min.js 
and put at visdom/static/js/plotly-plotly.min.js
download from https://unpkg.com/bootstrap@3.3.7/dist/css/bootstrap.min.css 
and put at visdom/static/css/bootstrap.min.css
download from https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.eot 
and put at visdom/static/fonts/glyphicons-halflings-regular.eot
download from https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.eot?#iefix 
and put at visdom/static/fonts/glyphicons-halflings-regular.eot?#iefix
download from https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.woff2 
and put at visdom/static/fonts/glyphicons-halflings-regular.woff2
download from https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.woff 
and put at visdom/static/fonts/glyphicons-halflings-regular.woff
download from https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.ttf 
and put at visdom/static/fonts/glyphicons-halflings-regular.ttf
download from https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.svg#glyphicons_halflingsregular 
and put at visdom/static/fonts/glyphicons-halflings-regular.svg#glyphicons_halflingsregular
```
If you run the server and it hangs on downloading, you can download the next missing file yourself (based on the last file in the list you can confirm having)

**Comment by huoliangyu** [comment]
It works for me. Thank you very much !

**Comment by ksnzh** [comment]
@JackUrb How to download `https://unpkg.com/classnames@2.2.5`? Is this a file?

**Comment by ksnzh** [comment]
@JackUrb Yes, it's a js file. And following your instructions, I solved this problem. Thank you very much.

**Comment by AndrewZhao** [comment]
@huoliangyu @JackUrb @ksnzh 
hello,this [link](https://cdn.rawgit.com/STRML/react-grid-layout/0.14.0/dist/react-grid-layout.min.js) seems have some errors. Can you share me the files on this link? Thanks very much!

**Comment by xiaofengqing** [comment]
@AndrewZhao  Can you share me all of the files in visdom/static? Thanks very much!
My Email address is 674868552@qq.com

**Comment by logic1988** [comment]
Thank @JackUrb . I changed the download link to a shell command with a text replacement: 
```
wget https://unpkg.com/jquery@3.1.1/dist/jquery.min.js -O visdom/static/js/jquery.min.js
wget https://unpkg.com/bootstrap@3.3.7/dist/js/bootstrap.min.js -O visdom/static/js/bootstrap.min.js
wget https://unpkg.com/react-resizable@1.4.6/css/styles.css -O visdom/static/css/react-resizable-styles.css
wget https://unpkg.com/react-grid-layout@0.14.0/css/styles.css
   and put at visdom/static/css/react-grid-layout-styles.css
wget https://unpkg.com/react@15.6.1/dist/react.min.js
   and put at visdom/static/js/react-react.min.js
wget https://unpkg.com/react-dom@15.6.1/dist/react-dom.min.js -O visdom/static/js/react-dom.min.js
wget https://unpkg.com/classnames@2.2.5 -O visdom/static/fonts/classnames
wget https://unpkg.com/layout-bin-packer@1.2.2 -O visdom/static/fonts/layout_bin_packer
wget https://cdn.rawgit.com/STRML/react-grid-layout/0.14.0/dist/react-grid-layout.min.js -O visdom/static/js/react-grid-layout.min.js
wget https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_SVG -O visdom/static/js/mathjax-MathJax.js


wget https://cdn.rawgit.com/plotly/plotly.js/master/dist/plotly.min.js -O visdom/static/js/plotly-plotly.min.js
wget https://unpkg.com/bootstrap@3.3.7/dist/css/bootstrap.min.css -O visdom/static/css/bootstrap.min.css
wget https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.eot -O visdom/static/fonts/glyphicons-halflings-regular.eot
wget https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.eot?#iefix -O visdom/static/fonts/glyphicons-halflings-regular.eot?#iefix
wget https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.woff2 -O visdom/static/fonts/glyphicons-halflings-regular.woff2
wget https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.woff -O visdom/static/fonts/glyphicons-halflings-regular.woff
wget https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.ttf -O visdom/static/fonts/glyphicons-halflings-regular.ttf
wget https://unpkg.com/bootstrap@3.3.7/dist/fonts/glyphicons-halflings-regular.svg#glyphicons_halflingsregular -O visdom/static/fonts/glyphicons-halflings-regular.svg#glyphicons_halflingsregular
```

**Comment by JackUrb** [comment]
@logic1988 thanks for the suggested conversion - I'll update the docs with this (as in the best case it automates the download and in the worst case if someone can't wget through their proxy it's no worse than the copy-paste it was before)

**Comment by sergeywong** [comment]
I meet the same problem. It is stuck at downloading the file `visdom/static/js/react-grid-layout.min.js ` from the link `https://cdn.rawgit.com/STRML/react-grid-layout/0.14.0/dist/react-grid-layout.min.js`. The problem occurs in the version 0.1.6 of `visdom`. The problem is resolved with  the version 0.1.5.

**Comment by JackUrb** [comment]
Hi @sergeywong - a fix for this was added to the github recently and I just pushed it to pypi. Try using version `0.1.6.4`
